### ISSUE TITLE
Plans to support Apple Silicon?

### ISSUE THREAD

**Issue by zbrox** [plotly/Kaleido#70]
Hey,

Plotly is much easier to use with Kaleido, but currently it doesn't seem possible to use it on one of Apple's new M1 machines. Is there a plan to support the Apple Silicon architecture?

**Comment by jonmmease** [comment]
Hi @zbrox, yes this is something we want to look into. Thanks for bringing it up!

 - Looks like we should be able to cross-compile for Applie Silicon on CircleCI: https://support.circleci.com/hc/en-us/articles/360047734172-Can-I-build-Apple-Silicon-Universal-apps-
 - Chromium build instructions: https://chromium.googlesource.com/chromium/src.git/+/master/docs/mac_arm64.md

Are you available to try out dev builds on an Apple Silicon machine when we have them ready to try?

**Comment by zbrox** [comment]
Yes, absolutely.

**Comment by jonmmease** [comment]
Great! I'll see if I can give it a try later this week.

**Comment by Endogen** [comment]
I'd be also interested in getting this working

**Comment by jonmmease** [comment]
FYI: Working on it in https://github.com/plotly/Kaleido/pull/71

But the required update to the base chromium version broke a bunch of CI and build configuration assumptions, so it'll take a bit more time to get everything straightened out and tested on the existing platforms.

I'll update this issue once there's a wheel ready to try out

**Comment by jonmmease** [comment]
Hi all, we have something ready to try out on apple silicon

### Testing
While kaleido can operate fully offline (and does so when used from plotly.py), for simplicity the first two tests pull plotly.js from an online CDN location, so they require internet access.  Test 3 (using kaleido from plotly.py) doesn't require internet access.

#### Test from command line
 1. From the dev release, download the [`kaleido_mac_arm64.zip`](https://github.com/plotly/Kaleido/releases/download/v0.1.0.post1.dev31%2Bg672b9fa/kaleido_mac_arm64.zip) file and unzip it.
 2. Use the terminal to cd into the unzipped `kaleido_mac_arm64` directory.
 3. Run this command:
 
```
$ ./kaleido plotly --no-sandbox --disable-gpu --disable-breakpad --allow-file-access-from-files
```

Please provide the full output printed to the terminal.

 4. If the above was successful, the kaleido process will be running and waiting for input. Paste this JSON bundle into the terminal and press enter.
 
```
{"data":{"layout":{}},"format":"svg"}
```

 5. Please provide the full output printed to the terminal. If successful, there should be a SVG string in there.


#### Test from Python
Install the pre-release kaleido macosx arm64 wheel from github

```
$ pip install -U plotly
$ pip install https://github.com/plotly/Kaleido/releases/download/v0.1.0.post1.dev31%2Bg672b9fa/kaleido-0.1.0.post1.dev31_g672b9fa.dev3_g81d6122-py2.py3-none-macosx_10_11_arm64.whl
```

If this fails, please provide all terminal output and your version of pip. If successful, launch a python or ipython repl and run this:

```python
from kaleido.scopes.plotly import PlotlyScope
scope = PlotlyScope()
scope.transform({"layout":{}}, format="svg")
```

If successful, this should print out an SVG image bytes object.  Try changing the `format` keyword from `svg `to `png`, `jpg`, `webm`, and `pdf` as well.

#### Test from plotly.py (no internet connection required)
After installing the pip packages above:

```python
import plotly.io as pio
print(pio.to_image({}, engine="kaleido", format="svg"))
```

If successful, this should print out an SVG image bytes object.  Try changing the `format` keyword from `svg `to `png`, `jpg`, `webm`, and `pdf` as well.

There are more tests we should do before release, but the results here will be very helpful to start with.

cc @zbrox @Endogen

**Comment by zbrox** [comment]
Hey! Thanks for the heads up. I'll try it right away. 

Here's the output of the CLI test first

```
[0224/153503.947196:WARNING:resource_bundle.cc(431)] locale_file_path.empty() for locale en-US
[0224/153504.403361:WARNING:dns_config_service_posix.cc(185)] dns_config has unhandled options!
[0224/153504.419441:WARNING:resource_bundle.cc(431)] locale_file_path.empty() for locale en-US
[0224/153504.427905:WARNING:resource_bundle.cc(431)] locale_file_path.empty() for locale en-US
[0224/153504.429683:WARNING:dns_config_service_posix.cc(185)] dns_config has unhandled options!
{"code": 0, "message": "Success", "result": null, "version": "0.1.0.post1.dev31+g672b9fa.dev3+g81d6122"}
[0224/153504.477317:WARNING:resource_bundle.cc(431)] locale_file_path.empty() for locale en-US
```

After input:

```json

{"code":0,"message":null,"pdfBgColor":null,"format":"svg","result":"<svg class=\"main-svg\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"700\" height=\"500\" style=\"\" viewBox=\"0 0 700 500\"><rect x=\"0\" y=\"0\" width=\"700\" height=\"500\" style=\"fill: rgb(255, 255, 255); fill-opacity: 1;\"/><defs id=\"defs-b12432\"><g class=\"clips\"><clipPath id=\"clipb12432xyplot\" class=\"plotclip\"><rect width=\"540\" height=\"320\"/></clipPath><clipPath class=\"axesclip\" id=\"clipb12432x\"><rect x=\"80\" y=\"0\" width=\"540\" height=\"500\"/></clipPath><clipPath class=\"axesclip\" id=\"clipb12432y\"><rect x=\"0\" y=\"100\" width=\"700\" height=\"320\"/></clipPath><clipPath class=\"axesclip\" id=\"clipb12432xy\"><rect x=\"80\" y=\"100\" width=\"540\" height=\"320\"/></clipPath></g><g class=\"gradients\"/></defs><g class=\"bglayer\"/><g class=\"layer-below\"><g class=\"imagelayer\"/><g class=\"shapelayer\"/></g><g class=\"cartesianlayer\"><g class=\"subplot xy\"><g class=\"layer-subplot\"><g class=\"shapelayer\"/><g class=\"imagelayer\"/></g><g class=\"gridlayer\"><g class=\"x\"><path class=\"xgrid crisp\" transform=\"translate(234.29,0)\" d=\"M0,100v320\" style=\"stroke: rgb(238, 238, 238); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"xgrid crisp\" transform=\"translate(311.43,0)\" d=\"M0,100v320\" style=\"stroke: rgb(238, 238, 238); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"xgrid crisp\" transform=\"translate(388.57,0)\" d=\"M0,100v320\" style=\"stroke: rgb(238, 238, 238); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"xgrid crisp\" transform=\"translate(465.71,0)\" d=\"M0,100v320\" style=\"stroke: rgb(238, 238, 238); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"xgrid crisp\" transform=\"translate(542.86,0)\" d=\"M0,100v320\" style=\"stroke: rgb(238, 238, 238); stroke-opacity: 1; stroke-width: 1px;\"/></g><g class=\"y\"><path class=\"ygrid crisp\" transform=\"translate(0,292)\" d=\"M80,0h540\" style=\"stroke: rgb(238, 238, 238); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"ygrid crisp\" transform=\"translate(0,228)\" d=\"M80,0h540\" style=\"stroke: rgb(238, 238, 238); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"ygrid crisp\" transform=\"translate(0,164)\" d=\"M80,0h540\" style=\"stroke: rgb(238, 238, 238); stroke-opacity: 1; stroke-width: 1px;\"/></g></g><g class=\"zerolinelayer\"><path class=\"xzl zl crisp\" transform=\"translate(157.14,0)\" d=\"M0,100v320\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"yzl zl crisp\" transform=\"translate(0,356)\" d=\"M80,0h540\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/></g><path class=\"xlines-below\"/><path class=\"ylines-below\"/><g class=\"overlines-below\"/><g class=\"xaxislayer-below\"/><g class=\"yaxislayer-below\"/><g class=\"overaxes-below\"/><g class=\"plot\" transform=\"translate(80,100)\" clip-path=\"url('#clipb12432xyplot')\"/><g class=\"overplot\"/><path class=\"xlines-above crisp\" d=\"M0,0\" style=\"fill: none;\"/><path class=\"ylines-above crisp\" d=\"M0,0\" style=\"fill: none;\"/><g class=\"overlines-above\"/><g class=\"xaxislayer-above\"><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"433\" transform=\"translate(80,0)\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(68, 68, 68); fill-opacity: 1; white-space: pre;\">−1</text></g><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"433\" transform=\"translate(157.14,0)\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(68, 68, 68); fill-opacity: 1; white-space: pre;\">0</text></g><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"433\" transform=\"translate(234.29,0)\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(68, 68, 68); fill-opacity: 1; white-space: pre;\">1</text></g><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"433\" transform=\"translate(311.43,0)\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(68, 68, 68); fill-opacity: 1; white-space: pre;\">2</text></g><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"433\" transform=\"translate(388.57,0)\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(68, 68, 68); fill-opacity: 1; white-space: pre;\">3</text></g><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"433\" transform=\"translate(465.71,0)\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(68, 68, 68); fill-opacity: 1; white-space: pre;\">4</text></g><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"433\" transform=\"translate(542.86,0)\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(68, 68, 68); fill-opacity: 1; white-space: pre;\">5</text></g><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"433\" transform=\"translate(620,0)\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(68, 68, 68); fill-opacity: 1; white-space: pre;\">6</text></g></g><g class=\"yaxislayer-above\"><g class=\"ytick\"><text text-anchor=\"end\" x=\"79\" y=\"4.199999999999999\" transform=\"translate(0,420)\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(68, 68, 68); fill-opacity: 1; white-space: pre;\">−1</text></g><g class=\"ytick\"><text text-anchor=\"end\" x=\"79\" y=\"4.199999999999999\" transform=\"translate(0,356)\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(68, 68, 68); fill-opacity: 1; white-space: pre;\">0</text></g><g class=\"ytick\"><text text-anchor=\"end\" x=\"79\" y=\"4.199999999999999\" transform=\"translate(0,292)\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(68, 68, 68); fill-opacity: 1; white-space: pre;\">1</text></g><g class=\"ytick\"><text text-anchor=\"end\" x=\"79\" y=\"4.199999999999999\" transform=\"translate(0,228)\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(68, 68, 68); fill-opacity: 1; white-space: pre;\">2</text></g><g class=\"ytick\"><text text-anchor=\"end\" x=\"79\" y=\"4.199999999999999\" transform=\"translate(0,164)\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(68, 68, 68); fill-opacity: 1; white-space: pre;\">3</text></g><g class=\"ytick\"><text text-anchor=\"end\" x=\"79\" y=\"4.199999999999999\" transform=\"translate(0,100)\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(68, 68, 68); fill-opacity: 1; white-space: pre;\">4</text></g></g><g class=\"overaxes-above\"/></g></g><g class=\"polarlayer\"/><g class=\"ternarylayer\"/><g class=\"geolayer\"/><g class=\"funnelarealayer\"/><g class=\"pielayer\"/><g class=\"treemaplayer\"/><g class=\"sunburstlayer\"/><g class=\"glimages\"/><defs id=\"topdefs-b12432\"><g class=\"clips\"/></defs><g class=\"layer-above\"><g class=\"imagelayer\"/><g class=\"shapelayer\"/></g><g class=\"infolayer\"><g class=\"g-gtitle\"/><g class=\"g-xtitle\"/><g class=\"g-ytitle\"/></g></svg>","width":700,"height":500,"scale":1}
```

**Comment by zbrox** [comment]
As for the python test, plotly install but not the kaleido wheel. I am using pip 21.0.1 and the error message is the following:

```
ERROR: kaleido-0.1.0.post1.dev31_g672b9fa.dev3_g81d6122-py2.py3-none-macosx_10_11_arm64.whl is not a supported wheel on this platform.
```

I double checked and I was using a venv with an arm64 python binary.

**Comment by zbrox** [comment]
This is what distutils reports as platform, if that can be helpful: `'macosx-11-arm64'`

**Comment by jonmmease** [comment]
Thanks @zbrox!  The result of the command-line test is super encouraging, it looks like the kaleido chromium build is working properly.

> This is what distutils reports as platform, if that can be helpful: 'macosx-11-arm64'

Ok, yeah. Looks like something with our platform tag isn't working. I've had trouble finding documentation for the apple silicon platform tag, but I'll kick off a new build with the `macosx-11-arm64` platform tag.

**Comment by zbrox** [comment]
@jonmmease whenever you have the release with that platform tag, just ping me and I'll test it. Thanks again!

**Comment by jonmmease** [comment]
@zbrox, the updated dev release is at https://github.com/plotly/Kaleido/releases/tag/v0.1.1a0.

The pip package is at https://github.com/plotly/Kaleido/releases/download/v0.1.1a0/kaleido-0.1.1a0-py2.py3-none-macosx_11_arm64.whl

Thanks!

**Comment by zbrox** [comment]
@jonmmease Unfortunately I still get the same error :/ 

```
ERROR: kaleido-0.1.1a0-py2.py3-none-macosx_11_arm64.whl is not a supported wheel on this platform.
```

**Comment by zbrox** [comment]
So, I had a hunch and decided to not trust what distutils was reporting. Downloaded the wheel and renamed it to add the macos minor version and it worked.

```
Processing ./kaleido-0.1.1a0-py2.py3-none-macosx_11_0_arm64.whl
Installing collected packages: kaleido
Successfully installed kaleido-0.1.1a0
```

**Comment by zbrox** [comment]
Both python tests ran successfully. I tried all formats you mentioned. I just used `webp` instead of `webm`, I guessed that was a typo. All seems good. Thank you!

**Comment by jonmmease** [comment]
Thanks for trying out adding the minor version, yeah that makes sense and I'll make that update.  And you're right, `webm` was a typo.

**Comment by jonmmease** [comment]
Ok, let's try this one more time! :crossed_fingers:  Release [0.1.1a1](https://github.com/plotly/Kaleido/releases/tag/v0.1.1a1)

Wheel:

https://github.com/plotly/Kaleido/releases/download/v0.1.1a1/kaleido-0.1.1a0.dev1+g5e23274-py2.py3-none-macosx_11_0_arm64.whl

**Comment by zbrox** [comment]
Yup! Looks good! 

<img width="1055" alt="image" src="https://user-images.githubusercontent.com/83163/109192880-12dade00-7798-11eb-96a7-747f9e23b2be.png">

**Comment by jonmmease** [comment]
Great, thanks @zbrox!

There a few more things I'd like to get into the next release, so if it's an option for you please go ahead and keep using this dev build and let us know if you run into anything.

**Comment by jonmmease** [comment]
Released in 0.2.0
